firmware/src/rdm_responder.c

```diff
@@ -149,411 +120,269 @@ static void SendDUBResponseIfRequired(const uint8_t *param_data,
       header, NACK_REASON,
       (header->command_class == GET_COMMAND ?
        GET_COMMAND_RESPONSE : SET_COMMAND_RESPONSE),
-      ntohs(header->param_id),
-      (uint8_t*) &param_data, sizeof(param_data));
+       ntohs(header->param_id),
+       sizeof(param_data));
+  memcpy(g_rdm_buffer + sizeof(RDMHeader), &param_data, sizeof(param_data));
+  return RDMUtil_AppendChecksum(g_rdm_buffer);
+}
+
+int RDMResponder_DispatchPID(const RDMHeader *header,
+                             const uint8_t *param_data) {
+  const ResponderDefinition *definition = g_responder.def;
+  uint16_t pid = ntohs(header->param_id);
+  unsigned int i = 0;
+  // TODO(simon): convert to binary search if the list gets long.
+  // We'll need to add a check to ensure it's sorted though.
+  for (; i < definition->descriptor_count; i++) {
+    if (pid == definition->descriptors[i].pid) {
+      if (header->command_class == GET_COMMAND) {
+        if (definition->descriptors[i].get_handler) {
+          return definition->descriptors[i].get_handler(header, param_data);
+        } else {
+          return RDMResponder_BuildNack(header, NR_UNSUPPORTED_COMMAND_CLASS);
+        }
+      } else {
+        if (definition->descriptors[i].set_handler) {
+          return definition->descriptors[i].set_handler(header, param_data);
+        } else {
+          return RDMResponder_BuildNack(header, NR_UNSUPPORTED_COMMAND_CLASS);
+        }
+      }
+    }
+  }
+  return RDMResponder_BuildNack(header, NR_UNKNOWN_PID);
 }
 
 // PID Handlers
 // ----------------------------------------------------------------------------
-
-static void Mute(const RDMHeader *header,
-                 unsigned int param_data_length) {
-  if (param_data_length != 0) {
-    return;
+int RDMResponder_GenericReturnString(const RDMHeader *header,
+                                     const char *reply_string,
+                                     unsigned int max_size) {
+  if (header->param_data_length) {
+    return RDMResponder_BuildNack(header, NR_FORMAT_ERROR);
   }
-  g_rdm_responder.is_muted = true;
-
-  uint16_t *control_field = (uint16_t*) g_rdm_responder.param_data;
-  *control_field = 0;
-  RespondIfRequired(header, ACK, DISCOVER_COMMAND_RESPONSE, PID_DISC_MUTE,
-                    g_rdm_responder.param_data, sizeof(*control_field));
-  PLIB_PORTS_PinClear(
-    PORTS_ID_0, g_rdm_responder.mute_port,
-    g_rdm_responder.mute_bit);
+  ReturnUnlessUnicast(header);
+
+  unsigned int length = RDMUtil_SafeStringLength(reply_string, max_size);
+  RDMResponder_BuildHeader(header, ACK, GET_COMMAND_RESPONSE,
+                           ntohs(header->param_id), length);
+  memcpy(g_rdm_buffer + sizeof(RDMHeader), reply_string, length);
+  return RDMUtil_AppendChecksum(g_rdm_buffer);
 }
 
-static void UnMute(const RDMHeader *header,
-                   unsigned int param_data_length) {
-  if (param_data_length != 0) {
-    return;
+int RDMResponder_SetMute(const RDMHeader *header) {
```

Comment:
I added a TODO, it's a bit tricky because I have a verification test that compares function pointers.